lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,186 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+#define MUTEX_MAX_REC_LOCK 32767
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+};
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
+{
+	int rc = 0;
+	int key = irq_lock();
+
+	if (m->owner == pthread_self()) {
+		if (m->type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (m->type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((m->owner == pthread_self()) || (m->lock_count == 0)) {
+			if (m->lock_count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			m->lock_count++;
```

Comment:
k_sem_take can fail, in which case it's not correct to increment the lock count, right?